overlord/fdestate/handlers.go

```diff
@@ -63,3 +67,81 @@ func (m *FDEManager) doRenameKeys(t *state.Task, tomb *tomb.Tomb) error {
 	//     reduces the possible states we could end up in.
 	return nil
 }
+
+func getCachedChangeAuthOptionsOnce(st *state.State) (*changeAuthOptions, error) {
+	cached := st.Cached(changeAuthOptionsKey{})
+	if cached == nil {
+		return nil, state.ErrNoState
+	}
+	st.Cache(changeAuthOptionsKey{}, nil)
+	var ok bool
+	opts, ok := cached.(*changeAuthOptions)
+	if !ok {
+		return nil, fmt.Errorf("internal error: wrong data type under changeAuthOptionsKey")
+	}
+	return opts, nil
+}
+
+func (m *FDEManager) doChangeAuthKeys(t *state.Task, _ *tomb.Tomb) error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	var keyslotRefs []KeyslotRef
+	if err := t.Get("keyslots", &keyslotRefs); err != nil {
+		return err
+	}
+
+	var authMode device.AuthMode
+	if err := t.Get("auth-mode", &authMode); err != nil {
+		return err
+	}
+
+	opts, err := getCachedChangeAuthOptionsOnce(m.state)
+	if errors.Is(err, state.ErrNoState) {
+		return fmt.Errorf("cannot find authentication options in memory: unexpected snapd restart")
+	}
+	if err != nil {
+		return fmt.Errorf("cannot find authentication options in memory: %v", err)
+	}
+
+	if opts.old == opts.new {
+		// no-op
+		return nil
```

Comment:
Why should we do this here instead of testing it in `ChangeAuth`?

Reply:
good point, this is to avoid breaking the API really as on success it expects an async response with a change ID, and if we have an empty change, it stays at `Hold` status forever, so it is more for convenience.